### PR TITLE
More GraphQL schema refactorings

### PR DESCRIPTION
**Pull request by andreas**
This PR contains a number of changes related to the schema in `irmin-graphql`. It could be split up in multiple PRs, but because the changes are dependent on each other (need to be applied in sequence), I've opted for a single, bigger PR. Hopefully reviewing the commits one by one can help make it more manageable. The commits should work independently if we want to omit one or more of them.

## Fix mixup of absolute and relative keys (3f3760e)

#663 introduced a bug (52f81f723635f5437167f148940952bda1e258ab):

> The source type of Node does not change (`tree * key`), but the semantics change: the tree is now fully resolved to the actual subtree, and the key is only part of the source type to support the `key` field.

Unfortunately the code mixed up relative and absolute keys: Keys used as arguments are relative to the current tree, but keys returned in responses are always absolute.

## More consistent fields on `Branch` and `Node` (8d373cf)

As mentioned in #663, I thought the "getters" on the GraphQL types `Branch` and `Node` should be consistent. The field `get` on `Branch` has type `String`, while the field `get` on `Node` has type `Tree`. Also, the two types offer different getters, e.g. `Branch` has `get_all: [Contents!]`, while `Node` does not.

I updated the getters on `Node` and `Branch` as follows in SDL terms:

```diff
type Node {
-  get(key: Key!): Node
+  get(key: Key!): String
+  get_all_contents: [Contents!]!
+  get_contents(key: Key!): Contents
+  get_node(key: Key!): Node
-  tree: [Tree!]!
+  list: [Tree!]!
  ...
}

type Branch {
  get(key: Key!): String
-  get_all: [Contents!]!
+  get_all_contents: [Contents!]!
+  get_node(key: Key!): Node
  ...
}
```

To elaborate:

- If the API consumer knows the key and kind of a particular key, you can use either `get`, `get_contents` or `get_node`.
- `get_all_contents` mirrors `Branch.get_all`, but I think the name `get_all_contents` is more clear.
- I renamed `Node.tree` to `Node.list` to reflect the type is not `Tree`.

I'm open to suggestions for naming!

## Replace getters on `Branch` with field `node` (88663f2)

After refactoring the getters on `Node`, the code duplication between the fields on `Branch` and `Node` annoyed me. The implementation were almost the same, but not quite. Further, any query using the getters on `Branch` could be rewritten to not using them, i.e. they were essentially convenience fields.

Example:

```grapqhl
query {
  master {
    get(key: "foo")
  }
}

query {
  master {
    head {
      node {
        get(key: "foo")
      }
    }
  }
}
```

To avoid the duplication, but still offer some convenience, I've removed them and added a convenience field `node: Node!` instead. This would allow the above example to be written as:

```diff
query {
  master {
    node {
      get(key: "foo")
    }
  }
}
```

Not quite as concise, but acceptable considering the upside?

## Swap names of `Tree` and `Node`(7742025)

This change is probably highly subjective. Prior to this change the terminology was as follows:

- `Contents` is a leaf node.
- `Node` is an internal node.
- `Tree` is the union of `Contents` and `Node`.

In my mind, `Node` is more general than `Tree`, so I suggest `Tree` to be an internal node, and `Node` to be the union of `Contents` and `Node`. After making the change, looking over the code and trying out some queries, I think it makes sense. Happy to hear what you think.

Note that the SDL diffs above are _without_ this change. Here's a full diff in SDL terms across all commits with this change:

```diff
 type Branch {
-  get(key: Key!): String
-  get_all(key: Key!): Contents
-  get_tree(key: Key!): [Contents!]
+  tree: Tree!
   head: Commit
   lcas(commit: CommitHash!): [Commit!]!
   name: String!
 }
 
 type Commit {
   hash: String!
   info: Info!
-  node: Node!
+  tree: Tree!
   parents: [String!]!
 }
 
-type Node {
+type Tree {
-  get(key: Key!): Tree
+  get(key: Key!): String
+  get_all_contents: [Contents!]!
+  get_contents(key: Key!): Contents
+  get_tree(key: Key!): Tree
  hash: String!
  key: String!
-  tree: [Tree!]!
+  list: [Node!]!
}

-union Tree = Contents | Node
+union Node = Contents | Tree
```

### CODE REVIEW

**Comment by zshipko** [comment]
Thanks for finding all of this and writing everything up!

Just one thing: before this PR `get_all` returned a single `Contents` value (like `Store.get_all`), this was only needed because `get` returned a `string` value so there was no way to get the associated metadata. Since all of that mess has been cleaned up and the new `get_all_contents` function isn't really related maybe it makes sense to give it a name closer to what Irmin would call it? Something like `list_contents`. What do you think?

Other than that, everything looks great!

**Comment by andreas** [comment]
> [...] the new `get_all_contents` function isn't really related maybe it makes sense to give it a name closer to what Irmin would call it? Something like `list_contents`

I agree `get_all_contents` is not great. I think the hard part is how to convey that the field traverses the entire subtree and pulls out all the content nodes. The field `list` only returns the immediate children, so I'm not sure it's intuitive that `list_contents` has different behavior.

Brainstorm:

- `list_all_contents` (`all` is supposed to convey different behavior 🤔)
- `list_contents_recursively`
- `list_all_contents_in_tree`
- `flatten_tree`

**Comment by zshipko** [comment]
Hmm, `flatten_tree` does a good job of conveying what it does without being too wordy. ‐ but `list_contents_recusively` is also good since it is a more exact description.

**Comment by andreas** [comment]
My votes goes to `list_contents_recusively` of the proposals so far. I think it's the most accurate, though it's quite lengthy. I'll update the PR..
